File: CPP/ByRound/1882/1882_B_Sets_and_Union.cpp
// Problem: CF 1882 B - Sets and Union
// https://codeforces.com/contest/1882/problem/B

/*
B. Sets and Union
Algorithms/Techniques: Bit manipulation, greedy approach
Time Complexity: O(n * max_element) where max_element <= 50
Space Complexity: O(n)

The problem asks to find the maximum number of elements in a set that can be formed
by taking union of some subsets from given sets, such that this set is NOT equal
to the union of ALL given sets.

Approach:
1. For each bit position (from 1 to 50), we check if it's present in the total union.
2. If yes, we count how many subsets do NOT contain this bit.
3. We form a mask of all elements from such subsets and calculate popcount.
4. The maximum among all such counts is our answer.

*/

#include <stdint.h>
#include <stdio.h>
#include <algorithm>

using namespace std;
#define LL long long
int T, n, k, x;
LL a[55], s; // a[i] stores the set as bitmask, s stores total union
int32_t main() {
  scanf("%d", &T);
  while (T--) {
    int mx = 0;
    s = 0;
    scanf("%d", &n);
    for (int i = 1; i <= n; i++) {
      a[i] = 0; // Clear the bitmask for current set
      scanf("%d", &k);
      while (k--) {
        scanf("%d", &x);
        a[i] |= 1ll << x; // Set the bit corresponding to element x
      }
      s |= a[i]; // Union all sets into s
    }
    for (int i = 1; i <= 50; i++)
      if (s >> i & 1) { // If bit i is set in total union
        LL tmp = 0;
        for (int j = 1; j <= n; j++)
          if (!(a[j] >> i & 1)) // If bit i is NOT set in subset j
            tmp |= a[j];       // Combine all elements from such subsets
        mx = max(mx, __builtin_popcountll(tmp)); // Count bits in tmp and update max
      }
    printf("%d\n", mx);
  }
}


// https://github.com/VaHiX/codeForces/

File: Python/ByTier/B/1882_B_Sets_and_Union.py
# Problem: CF 1882 B - Sets and Union
# https://codeforces.com/contest/1882/problem/B

"""
Purpose: This program finds the maximum number of elements in an attainable set S such that S is not equal to the union of all given sets S1, S2, ..., Sn.

Algorithm:
- For each element in the universal set (union of all sets), determine how many sets do NOT contain this element.
- Compute the union of all sets that do not contain the current element.
- The size of this union represents a possible attainable set that excludes at least one element from the total union.
- The maximum such size is the answer.

Time Complexity: O(n * m) where n is number of sets and m is average size of each set.
Space Complexity: O(m) where m is the total number of unique elements across all sets.

Note: A key insight is that we're looking for the maximum size of a union of subsets that is strictly less than the full union.
"""

n = int(input())
for _ in range(n):
    nn = int(input())
    k = [0] * nn
    s = [0 for __ in range(nn)]
    sall = set()  # Union of all sets
    for i in range(nn):
        k_l = list(map(int, input().split()))
        k[i] = k_l[0]
        s[i] = set(k_l[1:])
        sall = sall.union(s[i])
    lenall = len(sall)
    ans = 0
    fp = 0
    for x in sall:
        temps = set()  # Union of all sets that do NOT contain x
        for xs in s:
            if x not in xs:
                temps = temps.union(xs)
        lentemp = len(temps)
        if lentemp == lenall - 1:
            print(lenall - 1)
            fp = 1
            break
        ans = max(ans, lentemp)
    if fp == 0:
        print(ans)


# https://github.com/VaHiX/CodeForces/